### PR TITLE
[spirv] Port over cooperative matrix support

### PR DESCRIPTION
**Pull request by antiagainst**
This commit moves cooperative matrix support to the Linalg on
tensor path. A few improvements are done along the way to have
a better pipeline structure:

* Cooperative matrix support previoulsy were scattered across
  various passes; now they are scoped into the same pass for
  better isolation.
* Now the handling of cooperative matrix is at a more proper
  layer. Cooperative matrix is at a large vector semantics level.
  We try to lower vector transfer read/write/arthmetic ops when
  we are still at that level with the necessary high-level
  information; not trying to left them "fall through" to the
  final converting to SPIR-V stage.
* Now the lowering is more progressive. This commit leverages
  new constructs like unrealized_conversion_cast recently added
  in core to bridge the gap and makes SPIR-V lowering more modular
  and progressive: cooperative matrix part is first handled, with
  the rest latter converted and those cast ops cancelled out.
  Upstream SPIR-V support is expected to grow more towards this
  direction.

The majority of the code in the new VectorToCooperativeMatrixPass.cpp
file is moved from ConvertToSPIRVPass.cpp, with modifications to
properly use type conversion integrated into dialect conversion
framework.

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp:1]: 2021

File: iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp:47]: typo: matrix
```suggestion
  // Check if the matrix type can be supported as a cooperative matrix.
```

File: iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp:115]:
```suggestion
  // Returns true if the operation should be lowered using operations on
```

File: iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp:125]: typo
```suggestion
/// Convert subgroup level vector transfer to SPIR-V cooperative
```

File: iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/VectorToCooperativeMatrixPass.cpp:163]: Can we return failure() if it failed?

**Review by hanhanW — APPROVED** [review]
LGTM, just few nits! 👯

**Comment by hanhanW** [comment]
There are some TODO assigned to @ThomasRaoux so I think it's better to wait for his LGTM.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I am fine with moving this around, but we should just move this to core. The issue here is that the co-operative matrix has its own types and load-store instructions. So the vector transfer-reads and transfer-writes become problematic. THe best way to handle this is to update the Vector to Co-operative matrix lowering in core to not require the pre-analysis pass done here.  For this we are first missing a "spv.cast" operation that converts from the default `spv.array` type to the co-operative matrix type.

1) Lower the `vector.contract` using `spv.cast` -> `spv.cooperativeMatrixMulOp` -> `spv.cast`.
2) Then you have a `spv.load` -> `spv.cast` that should be converted to `spv.loadcooperativematrix` (and same for the store).

This then requires no state and this pass can be dropped. Would be better for SPIR-V lowering in general.

This refactoring by itself is good, but deleting this pass completely is even better (we have code cleanup as part of our OKRs this quarter)

File: iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp:219]: I don't understand why it's scary? This is just a simple canonicalization pattern, right? Unrealized conversion casts are expected to be the "bridges" between progressive lowerings and it should go away at the very last step when we convert everything to the final target, which is in this pass. Unrealized conversion casts here were inserted by dialect conversion framework to "cast" `hal.interface.binding.subspan` ops into its SPIR-V type, which won't be materialized until in this pass where we convert `hal.interface.binding.subspan` into SPIR-V global variables.

**Comment by antiagainst** [comment]
> I am fine with moving this around, but we should just move this to core. The issue here is that the co-operative matrix has its own types and load-store instructions. So the vector transfer-reads and transfer-writes become problematic. THe best way to handle this is to update the Vector to Co-operative matrix lowering in core to not require the pre-analysis pass done here. For this we are first missing a "spv.cast" operation that converts from the default `spv.array` type to the co-operative matrix type.
> 
> 1. Lower the `vector.contract` using `spv.cast` -> `spv.cooperativeMatrixMulOp` -> `spv.cast`.
> 2. Then you have a `spv.load` -> `spv.cast` that should be converted to `spv.loadcooperativematrix` (and same for the store).
> 
> This then requires no state and this pass can be dropped. Would be better for SPIR-V lowering in general.
> 
> This refactoring by itself is good, but deleting this pass completely is even better (we have code cleanup as part of our OKRs this quarter)

Yes right. The existing analysis is hacky. @ThomasRaoux is pushing for a more generic cooperative matrix ish representation in gpu dialect so that we can share between cuda and spirv land. After that's landed we can go from vector to gpu counterparts and then convert to spirv will be simple. Expected this part to be simplified in the future; right now doing this to unblock deprecating the buffer path.

**Review by ThomasRaoux — APPROVED** [review]
Thanks Lei!

File: iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp:219]: Ill approve this PR since it contains things better which is good, but to me `UnrealizedConversionCast` seems like a catch all, and an indication that something "wasnt handled". It is very easy for a lot of things to end up there, and very easy for something to work even though it was not suppsoed to work. Let me ask a different question, if the lowering of `vector.contract` to `spv.CoooperativeMatrix*` was done in this pass and you had the `hal.interface.binding.subspan` to SPIR-V type patterns, these instructions wouldnt exist right? Its helping you break the lowering of vector.contract to cop-operative matrix instructions and general lowering to SPIR-V, but the real issue is that the SPIR-V lowering to co-operative matrix multiply instructions need to be fixed in core to handle the different types used for these ops.

> > I am fine with moving this around, but we should just move this to core. The issue here is that the co-operative matrix has its own types and load-store instructions. So the vector transfer-reads and transfer-writes become problematic. THe best way to handle this is to update the Vector to Co-operative matrix lowering in core to not require the pre-analysis pass done here. For this we are first missing a "spv.cast" operation that converts from the default `spv.array` type to the co-operative matrix type.
> > 
> > 1. Lower the `vector.contract` using `spv.cast` -> `spv.cooperativeMatrixMulOp` -> `spv.cast`.
> > 2. Then you have a `spv.load` -> `spv.cast` that should be converted to `spv.loadcooperativematrix` (and same for the store).
> > 
> > This then requires no state and this pass can be dropped. Would be better for SPIR-V lowering in general.
> > This refactoring by itself is good, but deleting this pass completely is even better (we have code cleanup as part of our OKRs this quarter)
> 
> Yes right. The existing analysis is hacky. @ThomasRaoux is pushing for a more generic cooperative matrix ish representation in gpu dialect so that we can share between cuda and spirv land. After that's landed we can go from vector to gpu counterparts and then convert to spirv will be simple. Expected this part to be simplified in the future; right now doing this to unblock deprecating the buffer path.

@ThomasRaoux we need to talk about this. Use of GPU dialect might just displace the issue. The real fix needs to be someway in which SPIR-V itself can handle the special type for co-operative matrix multiplies.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp:219]: Thanks. There are multiple issues; type conversion being one of them. There certainly should be direct support in the SPIR-V type converter upstream to convert into cooperative matrix types (essentially the additional conversion rules I attached in this pass, with proper capability check, etc.). Having all of those properly implemented takes time. I mentioned Thomas' plans just to show that he is working on pushing for better solutions already, so we can expect this to be improved. Not saying going gpu is directly addressing all the issues; that was more for sharing with cuda side. Hope this clarifies. :)

Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp:219]: > @ThomasRaoux we need to talk about this. Use of GPU dialect might just displace the issue. The real fix needs to be someway in which SPIR-V itself can handle the special type for co-operative matrix multiplies.

Sure, let's discuss this. I think the GPU dialect should address most of the problem since it will give us an op working on memref analogue to std.load and we can support the new type in the SPIR-V type converter.

**Pull request merged**
